dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/programrule/implementers/SetMandatoryFieldValidator.java

```diff
@@ -118,10 +118,12 @@ private List<ProgramRuleIssue> checkMandatoryDataElement( Event event, List<Even
     {
         ProgramStage programStage = bundle.getPreheat().getProgramStage( event.getProgramStage() );
 
+        // TODO() what is this String, always a UID?
         Map<String, EventActionRule> mandatoryDataElementsByActionRule = actionRules.stream()
```

Comment:
@stian-sandvold and @enricocolasante I need your input on this

I want to make the "String" key a MetadataIdentifier since in `validateMandatoryDataValue` we compare against DataElements from the payload which can be in any idScheme.

What I dont know is what the String represents. Always a UID?
If so same as above. I would need to fetch that from the preheat so I can map it to a MetadataIdentifier. In that case the DataElement from the EventActionRule might not be found in the prehehat.

Reply:
I really think here the key is always as UID as in rule engine we do not have the idScheme concept.
So if a user creates a rule involving a data element it must always use the UID to identify it

Reply:
And do we ensure the user can only create a rule with data element that exists in the DB (thus tracker preheat)? Because I need to validate based on the idScheme in the payload, thus fetch the rule data element from the preheat

Reply:
I think we are sure that this is true when they create the rule. I am not sure we ensure that this will be true forever, let's say if someone delete the data element, I don't think we are deleting/invalidating the rule.

But @zubaira or @Bekkalizer should know better about it

Reply:
Hey! Rule actions have a foreign key constraint for both data elements, tracked entity attributes and other metadata object types. So if there is a data element linked to the rule action, it would always exist in the database. We would not be certain that the user can read it(with the metadata:read sharing level), but it would always be in the database.